### PR TITLE
*: merge IntColumnRange with NewRange.

### PR DESCRIPTION
**Pull request by winoros**
Remove `IntColumnRange` and use `NewRange` to store table range.
So uint64's condition can be handled correctly, but uint64's table range can not be send to tikv correctly. This will be fixed in next pr.

### CODE REVIEW

**Comment by winoros** [comment]
PTAL @hanfei1991 @coocood @lamxTyler

File: statistics/statistics_test.go [hunk 1]
```diff
@@ -524,47 +524,47 @@ func (s *testStatisticsSuite) TestIntColumnRanges(c *C) {
...
-	ran[0].LowVal = 1000
-	ran[0].HighVal = 2000
+	ran[0].LowVal[0].SetInt64(1000)
+	ran[0].HighVal[0].SetInt64(2000)
 	count, err = tbl.GetRowCountByIntColumnRanges(sc, 0, ran)
 	c.Assert(err, IsNil)
-	c.Assert(int(count), Equals, 1000)
-	ran[0].LowVal = 1001
-	ran[0].HighVal = 1999
+	c.Assert(int(count), Equals, 1001)
```
Comment (winoros) [inline statistics/statistics_test.go]: This is more accurate than before. ([1000, 2000] have 1001 numbers.)

File: statistics/statistics_test.go [hunk 2]
```diff
@@ -524,47 +524,47 @@ func (s *testStatisticsSuite) TestIntColumnRanges(c *C) {
...
+	ran[0].LowVal[0].SetInt64(1001)
+	ran[0].HighVal[0].SetInt64(1999)
 	count, err = tbl.GetRowCountByIntColumnRanges(sc, 0, ran)
 	c.Assert(err, IsNil)
-	c.Assert(int(count), Equals, 998)
-	ran[0].LowVal = 1000
-	ran[0].HighVal = 1000
+	c.Assert(int(count), Equals, 999)
```
Comment (winoros) [inline statistics/statistics_test.go]: [1001, 1999] have 999 numbers.

File: executor/distsql.go [hunk 3]
```diff
@@ -83,20 +83,23 @@ func (task *lookupTableTask) getRow(schema *expression.Schema) (Row, error) {
...
-	krs := make([]kv.KeyRange, 0, len(tableRanges))
-	for _, tableRange := range tableRanges {
-		startKey := tablecodec.EncodeRowKeyWithHandle(tid, tableRange.LowVal)
-
-		var endKey kv.Key
-		if tableRange.HighVal != math.MaxInt64 {
-			endKey = tablecodec.EncodeRowKeyWithHandle(tid, tableRange.HighVal+1)
-		} else {
-			endKey = tablecodec.EncodeRowKeyWithHandle(tid, tableRange.HighVal).Next()
+func tableRangesToKVRanges(tid int64, ranges []*ranger.NewRange) ([]kv.KeyRange, error) {
```
Comment (lamxTyler) [inline executor/distsql.go]: Why return `error` since there is no error to return?

File: statistics/update.go [hunk 4]
```diff
@@ -179,7 +179,7 @@ type QueryFeedback struct {
 	colID       int64
 	isIndex     bool
 	idxRanges   []*ranger.NewRange
-	intRanges   []ranger.IntColumnRange
+	intRanges   []*ranger.NewRange
```
Comment (lamxTyler) [inline statistics/update.go]: Only one of the `intRanges` and `idxRanges` will be set, so just keep one is fine.

File: statistics/table.go [hunk 5]
```diff
@@ -414,25 +420,57 @@ func getPseudoRowCountByColumnRanges(sc *stmtctx.StatementContext, tableRowCount
 	return rowCount, nil
 }
 
-func getPseudoRowCountByIntRanges(intRanges []ranger.IntColumnRange, tableRowCount float64) float64 {
+func getPseudoRowCountBySignedIntRanges(intRanges []*ranger.NewRange, tableRowCount float64) float64 {
```
Comment (jackysp) [inline statistics/table.go]: This function is complex and similar with uint version. Is there a good way to simplify it?

File: util/ranger/points.go [hunk 6]
```diff
@@ -128,6 +128,12 @@ func FullIntRange() []IntColumnRange {
 	return []IntColumnRange{{LowVal: math.MinInt64, HighVal: math.MaxInt64}}
 }
 
+// FullIntNewRange is used for table range. Since table range cannot accept MaxValueDatum as the max value.
+// So we need to set it to MaxInt64.
+func FullIntNewRange() []*NewRange {
+	return []*NewRange{{LowVal: []types.Datum{types.NewIntDatum(math.MinInt64)}, HighVal: []types.Datum{types.NewIntDatum(math.MaxInt64)}}}
```
Comment (hanfei1991) [inline util/ranger/points.go]: It's better set exculde flag of highVal be true, which can avoid a `prefixNext`

File: util/ranger/types.go [hunk 7]
```diff
@@ -134,12 +134,29 @@ func (ran *NewRange) PrefixEqualLen(sc *stmtctx.StatementContext) (int, error) {
...
+			if isLeftSide {
+				return "-inf"
+			}
+		case math.MaxInt64:
+			if !isLeftSide {
+				return "+inf"
+			}
+		}
+	case types.KindUint64:
+		if d.GetUint64() == math.MaxUint64 && !isLeftSide {
```
Comment (hanfei1991) [inline util/ranger/types.go]: When is it MaxUint64 but is on the left side.

Comment (winoros) [reply util/ranger/types.go]: `a >= MaxUint64` the actual range will be `[MaxUint64, MaxUint64]`. The best result to display is `[MaxUint64, +inf]` i think

File: util/ranger/points.go (hunk 6 above)
Comment (winoros) [reply util/ranger/points.go]: We cannot set it since this will not contain `MaxInt64`.

**Comment by hanfei1991** [comment]
LGTM

File: util/ranger/ranger.go [hunk 8]
```diff
@@ -161,65 +165,62 @@ func appendPoints2IndexRange(sc *stmtctx.StatementContext, origin *NewRange, ran
...
-		if cmp > 0 || (cmp == 0 && endPoint.excl) {
-			if endInt == math.MinInt64 {
-				continue
-			}
-			endInt--
+		if !less {
+			continue
 		}
-		if startInt > endInt {
+		if endPoint.value.Kind() == types.KindNull {
```
Comment (lamxTyler) [inline util/ranger/ranger.go]: Why not check it before `pointLess`?

File: statistics/table.go (hunk 5 above)
Comment (winoros) [reply statistics/table.go]: We can simplify this later. Datum's Compare method needs to handle error, it's not so good to use datum to merge this two function.

**Comment by winoros** [comment]
@lamxTyler PTAL

**Pull request merged**